### PR TITLE
ARM/ARM64 split - part 1

### LINKED ISSUE #32741: ARM32 / ARM64 separation
ARM32 and ARM64 (AArch64) currently share the same code space: `arch/arm`,`include/arch/arm`, ` soc/arm` and `boards/arm`.

This was done at the beginning of the time because the hope was that the two architectures would have shared more code. Reality is that ARM32 and AArch64 are two different architectures which shares almost nothing. The only thing they share is the GIC driver that is hosted outside of `arch/arm`.

So my proposal is that we should split the two and treat them as two different architectures. We could start separating them in `soc/{arm,arm64}` and `boards/{arm,arm64}` but I think that we should push this forward and moving to have a separate `arch/arm64` directory as well.

Any thoughts?

### PR DESCRIPTION
**Pull request by carlocaione**
Some important points:

- To have this passing on CI we need to have https://github.com/zephyrproject-rtos/sdk-ng/pull/325 merged and actively used by CI
- This is going to fail on checkpatch because it is going to recheck all the moved code so we are catching all the past warnings, etc...
- This is going to conflict often
- This PR is only for the AArch64 part. When this is merged I will post another PR taking care of the AArch32 part.
- It was a bit more painful than expected but I also think this is necessary

Closes #32741

### CODE REVIEW

**Review by galak — COMMENTED** [review]
Is the `aarch32` removal needed in this PR?  Agree it should be done, but wondering about splitting it out into its own PR?

**Comment by carlocaione** [comment]
> Is the `aarch32` removal needed in this PR? Agree it should be done, but wondering about splitting it out into its own PR?

Not strictly needed (this is why I have it in a separate commit). But I can split the PRs for the sake of simplicity.

**Comment by galak** [comment]
> Not strictly needed (this is why I have it in a separate commit). But I can split the PRs for the sake of simplicity.

Just thinking to get arm64 split merged to limit this PR as much as possible, and than have followup cleanup PRs would make it quicker to get merged.

**Comment by carlocaione** [comment]
> Just thinking to get arm64 split merged to limit this PR as much as possible, and than have followup cleanup PRs would make it quicker to get merged.

No objections. I'll use this PR only for the AArch64 part and once this is merged I can push the remaining part.

**Comment by npitre** [comment]
> No objections. I'll use this PR only for the AArch64 part and once this is merged I can push the remaining part.

+1

File: CODEOWNERS [hunk 1]
```diff
@@ -105,7 +104,7 @@
 /boards/arm/nucleo*/                      @erwango @ABOSTM @FRASTM
 /boards/arm/nucleo_f401re/                @idlethread
 /boards/arm/nuvoton_pfm_m487/             @ssekar15
-/boards/arm/qemu_cortex_a53/              @carlocaione
+/boards/arm64/                            @carlocaione
```
Comment (galak) [inline CODEOWNERS:107]: We should have this be broken up by board.  Since we only have a small handful its easier to do this now.

**Comment by carlocaione** [comment]
Alright, I split the PRs. This PR is now only concerning the AArch64 separation.

File: arch/arm64/Kconfig [hunk 2]
```diff
@@ -0,0 +1,39 @@
+# ARM64 architecture configuration options
+
+# Copyright (c) 2014-2015 Wind River Systems, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+menu "ARM Options"
```
Comment (galak) [inline arch/arm64/Kconfig:6]: "ARM 64 Options"

File: arch/arm64/Kconfig [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	select 64BIT
+
+config CPU_CORTEX
+	bool
+	help
+	  This option signifies the use of a CPU of the Cortex family.
+
+config ARM_CUSTOM_INTERRUPT_CONTROLLER
+	bool
+	depends on !CPU_CORTEX_M
```
Comment (galak) [inline arch/arm64/Kconfig:23]: Is the !CPU_CORTEX_M still needed?

File: arch/arm64/Kconfig [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	help
+	  This option indicates that the ARM CPU is connected to a custom (i.e.
+	  non-GIC) interrupt controller.
+
+	  A number of Cortex-A and Cortex-R cores (Cortex-A5, Cortex-R4/5, ...)
+	  allow interfacing to a custom external interrupt controller and this
```
Comment (galak) [inline arch/arm64/Kconfig:29]: Is the help here still relevant for arm64?

File: arch/arm64/core/Kconfig [hunk 5]
```diff
@@ -28,6 +25,12 @@ config CPU_CORTEX_A72
 	help
 	  This option signifies the use of a Cortex-A72 CPU
 
+config HAS_ARM_SMCCC
+	bool
+	help
+	  Include support for the Secure Monitor Call (SMC) and Hypervisor
+	  Call (HVC) instructions on Armv7 and above architectures.
```
Comment (galak) [inline arch/arm64/core/Kconfig:32]: armv7?

File: arch/arm64/Kconfig (hunk 2 above)
Comment (carlocaione) [reply arch/arm64/Kconfig:6]: ACK

File: arch/arm64/core/tls.c [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ */
+
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <kernel_internal.h>
+#include <kernel_tls.h>
+#include <app_memory/app_memdomain.h>
+#include <sys/util.h>
+
+#ifdef CONFIG_CPU_CORTEX_M
```
Comment (galak) [inline arch/arm64/core/tls.c:14]: Remove CONFIG_CPU_CORTEX_M

File: arch/arm64/Kconfig (hunk 3 above)
Comment (carlocaione) [reply arch/arm64/Kconfig:23]: ACK

File: arch/arm64/include/kernel_arch_data.h [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+#include <sys/atomic.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct __esf _esf_t;
+typedef struct __basic_sf _basic_sf_t;
+
+#ifdef CONFIG_ARM_MPU
```
Comment (galak) [inline arch/arm64/include/kernel_arch_data.h:42]: is ARM_MPU relevant on arm64?

File: arch/arm64/core/Kconfig (hunk 5 above)
Comment (carlocaione) [reply arch/arm64/core/Kconfig:32]: Yes, this is valid from armv7 up.

File: arch/arm64/Kconfig (hunk 4 above)
Comment (carlocaione) [reply arch/arm64/Kconfig:29]: yes, it is.

File: arch/arm64/core/tls.c (hunk 6 above)
Comment (carlocaione) [reply arch/arm64/core/tls.c:14]: ACK

File: arch/arm64/include/kernel_arch_data.h (hunk 7 above)
Comment (carlocaione) [reply arch/arm64/include/kernel_arch_data.h:42]: ACK

File: dts/arm64/broadcom/viper-m7.dtsi [hunk 8]
```diff
@@ -0,0 +1,54 @@
+/*
```
Comment (galak) [inline dts/arm64/broadcom/viper-m7.dtsi:1]: Should this move, its cortex-m7?

Comment (carlocaione) [reply dts/arm64/broadcom/viper-m7.dtsi:1]: ah! good catch. The viper board is shipping an a72 and an m7 so we have some DTS duplicated. I copied the directory but forgot to clean it up.

**Review by galak — CHANGES_REQUESTED** [review]
Some questions and a few cleanups in a few places.

File: CODEOWNERS (hunk 1 above)
Comment (carlocaione) [reply CODEOWNERS:107]: OK

**Comment by galak** [comment]
It feels like the broadcom SoC/board stuff might be useful to have in its own commit since I feel like the split going on there between arm64/arm isn't a straight forward move.

File: arch/arm64/core/Kconfig (hunk 5 above)
Comment (galak) [reply arch/arm64/core/Kconfig:32]: I thought arm64 started on armv8, but maybe my memory is wrong.

**Comment by carlocaione** [comment]
> It feels like the broadcom SoC/board stuff might be useful to have in its own commit since I feel like the split going on there between arm64/arm isn't a straight forward move.

That is going to break bisectability for the Broadcom board though.

**Comment by galak** [comment]
> That is going to break bisectability for the Broadcom board though.

At least maybe call out what's happening in the commit message for the broadcom board/soc.

File: arch/arm64/core/Kconfig (hunk 5 above)
Comment (carlocaione) [reply arch/arm64/core/Kconfig:32]: What I mean is that SMCCC is a standard valid also for armv7, hence the description in the help. But in Zephyr ARMv8 is the only to actually use that.

**Comment by carlocaione** [comment]
@Abhishek-brcm @sandeepbrcm about the bcm_vk/viper split it would be nice to have your review as well.

**Comment by carlocaione** [comment]
@galak there is a way to use zephyrproject-rtos/sdk-ng#325 for CI already?

**Review by npitre — APPROVED** [review]

**Comment by carlocaione** [comment]
The compliance test is failing as usual on the typedefs, so that's bogus as usual.

**Review by galak — APPROVED** [review]

**Pull request merged**